File: sdk/java/src/main/java/org/nanocached/NanocachedException.java
package org.nanocached;

/** Base class for every error this SDK raises on its own behalf. */
public class NanocachedException extends RuntimeException {
    public NanocachedException(String message) {
        super(message);
    }

    public NanocachedException(String message, Throwable cause) {
        super(message, cause);
    }

    /** Raised by get/set/delete after close(); close() itself is idempotent. */
    public static final class AlreadyClosed extends NanocachedException {
        public AlreadyClosed() {
            super("nanocached: this client is closed");
        }
    }

    /**
     * A node answered {@code W} (staged node join): per its own view of cluster
     * membership it doesn't hold this key — the caller's routing table is
     * stale. The client catches this internally to refresh the node list
     * and retry once; it only escapes when that retry also fails, or in
     * single-node mode where there is no discovery to refresh from.
     */
    public static class WrongNode extends NanocachedException {
        public WrongNode() {
            super("nanocached: this node does not hold the requested key");
        }
    }

    /**
     * The server rejected the {@code A} handshake's secret — either no
     * {@code authSecret} was configured for a server that requires one,
     * or the configured secret is wrong. Never transient: retrying with
     * the same configuration cannot succeed.
     */
    public static final class AuthenticationFailed extends NanocachedException {
        public AuthenticationFailed(String message) {
            super(message);
        }
    }

    /**
     * A discovery server answered {@code L} with {@code B} — it is inside
     * its startup grace (discovery HA), re-learning membership after a
     * restart. Try another address, or retry shortly.
     */
    public static final class DiscoveryBusy extends NanocachedException {
        public DiscoveryBusy() {
            super("nanocached: the discovery server is busy: warming up after a restart, or its replication factor disagrees with the cluster's");
        }
    }

    /** A connection-level failure; the client redials lazily on the next use. */
    public static final class ConnectionFailed extends NanocachedException {
        /**
         * Issue #225: {@code true} only when this connection is known to
         * have never sent the failed request's bytes at all — {@link
         * Connection#request}'s pre-write {@code isClosed()} checks are
         * the only place this is set, and it means the connection was
         * already marked closed before this call ever attempted to write,
         * almost always because the reader thread had already noticed the
         * peer's FIN (e.g. its idle timeout) moments earlier. Only then is
         * redialing and resending a <em>non-idempotent</em> request
         * (INCR/CAS/delete-if-matches) safe — see {@link
         * NanocachedClient}'s {@code applyReconnectingNonIdempotent}.
         * Every other {@code ConnectionFailed} (the write itself failing
         * partway, a request timeout, or the reply simply never arriving
         * after a successful write) leaves the request's fate genuinely
         * unknown, so it must not be replayed. Irrelevant to get/set/
         * delete/clear, which stay safe to retry unconditionally because
         * they're idempotent regardless of this flag — always {@code
         * false} on the public two-argument constructor those (and every
         * caller outside {@link Connection}) use.
         */
        private final boolean notSent;

        public ConnectionFailed(String message, Throwable cause) {
            this(message, cause, false);
        }

        ConnectionFailed(String message, Throwable cause, boolean notSent) {
            super(message, cause);
            this.notSent = notSent;
        }

        /** See {@link #notSent}. */
        public boolean notSent() {
            return notSent;
        }
    }

    /**
     * Raised by get/getBytes when a value with {@code compress} enabled
     * can't be interpreted — almost always a {@code compress} mismatch
     * between clients sharing this key (value compression's compatibility
     * caveat: every client touching a given keyspace must agree on
     * {@code compress}), not a transient failure.
     */
    public static final class DecompressionFailed extends NanocachedException {
        public DecompressionFailed(String message) {
            super(message);
        }

        public DecompressionFailed(String message, Throwable cause) {
            super(message, cause);
        }
    }

    /**
     * A request answered {@code R} (issue #125's retryable-error status)
     * three times running — bounded transparent retry on the SAME
     * connection (up to 2 retries, 3 attempts total, 50ms then 100ms
     * between them) exhausted without a real answer. Today only
     * {@code nanocached-proxy} sends {@code R}, when its upstream node was
     * briefly unreachable and survived its own one refresh-and-retry; this
     * is the signal that even that didn't resolve in time.
     *
     * <p>Unlike {@link ConnectionFailed} or {@link WrongNode}, this never
     * closes or redials the connection — it stays open and usable for the
     * next call. Every {@code R} answer, including the three that led
     * here, is counted in {@link NanocachedClient.ClientStats#transientRetries()}.
     */
    public static final class RetryableError extends NanocachedException {
        public RetryableError() {
            super("nanocached: request failed transiently (R) after 3 attempts on this connection");
        }
    }

    /**
     * A node answered {@code T} to an {@code i} (INCR/DECR, issue #129)
     * request: the key exists but its stored value isn't INCR's counter
     * grammar (plain decimal ASCII), or applying the delta would overflow
     * a 64-bit counter. Never transient — retrying the same request
     * answers the same way until the key is overwritten with a numeric
     * value (or deleted).
     */
    public static final class NotNumeric extends NanocachedException {
        public NotNumeric() {
            super("nanocached: the stored value is not an integer INCR can operate on");
        }
    }

    /**
     * Raised by {@link NanocachedClient#getManyBytes} (issue #151) when
     * some keys are still wrong-node after the one bounded
     * refresh-and-retry every batch gets (the per-key analogue of {@code
     * getBytes}' own {@code W} refresh-and-retry) — a subclass of {@link
     * WrongNode}, so existing {@code catch (WrongNode)} handling keeps
     * working unchanged. {@code partialValues} holds every key that DID
     * resolve — a batch never fails as a whole
     * (docs/protocol.html#multi), so a handful of stale placements
     * shouldn't force discarding an otherwise successful batch. {@link
     * NanocachedClient#setManyBytes} has nothing to return on success, so
     * it just throws a plain {@link WrongNode} on the same condition —
     * there's no partial payload worth attaching.
     *
     * <p>A separate, non-generic class from {@link PartialWrongNodeStrings}
     * rather than one class parameterized over the value type: the JLS
     * forbids a generic class from being a (direct or indirect) subclass
     * of {@link Throwable}.
     */
    public static final class PartialWrongNode extends WrongNode {
        public final java.util.Map<String, byte[]> partialValues;

        public PartialWrongNode(java.util.Map<String, byte[]> partialValues) {
            this.partialValues = partialValues;
        }
    }

    /**
     * As {@link PartialWrongNode}, but raised by {@link
     * NanocachedClient#getMany} — the UTF-8-decoded counterpart, thrown
     * once the raw {@link PartialWrongNode} its {@code getManyBytes} call
     * underneath threw has had its {@code partialValues} decoded the same
     * way a successful {@code getMany} decodes {@code getManyBytes}' own
     * result.
     */
    public static final class PartialWrongNodeStrings extends WrongNode {
        public final java.util.Map<String, String> partialValues;

        public PartialWrongNodeStrings(java.util.Map<String, String> partialValues) {
            this.partialValues = partialValues;
        }
    }

    /**
     * As {@link PartialWrongNode}, but raised by the positional,
     * {@code byte[]}-keyed {@link NanocachedClient#getManyBytes(byte[][])}
     * (issue #160). {@code partialValues} is the same positional array a
     * successful call would have returned ({@code null} for a miss);
     * since a {@code null} slot alone cannot tell a miss from a key that
     * is still wrong-node, {@code unresolvedIndices} lists (ascending)
     * the positions that did NOT resolve.
     */
    public static final class PartialWrongNodeRaw extends WrongNode {
        public final byte[][] partialValues;
        public final int[] unresolvedIndices;

        public PartialWrongNodeRaw(byte[][] partialValues, java.util.List<Integer> unresolvedIndices) {
            this.partialValues = partialValues;
            this.unresolvedIndices = unresolvedIndices.stream().mapToInt(Integer::intValue).sorted().toArray();
        }
    }
}

File: sdk/dotnet/src/Nanocached/NanocachedException.cs
namespace Nanocached;

/// <summary>Base class for every error this SDK raises on its own behalf.</summary>
public class NanocachedException : Exception
{
    public NanocachedException(string message) : base(message) { }

    public NanocachedException(string message, Exception inner) : base(message, inner) { }
}

/// <summary>Raised by get/set/delete after Close(); Close() itself is idempotent.</summary>
public sealed class AlreadyClosedException : NanocachedException
{
    public AlreadyClosedException() : base("nanocached: this client is closed") { }
}

/// <summary>
/// A node answered <c>W</c> (staged node join): per its own view of cluster
/// membership it doesn't hold this key — the caller's routing table is
/// stale. The client catches this internally to refresh the node list and
/// retry once; it only escapes when that retry also fails, or in
/// single-node mode where there is no discovery to refresh from.
///
/// <para>Not <c>sealed</c> (unlike this SDK's other leaf exceptions) so
/// <see cref="PartialWrongNodeException{T}"/> — issue #151's batched-get
/// partial-failure carrier — can subclass it while still satisfying every
/// existing <c>catch (WrongNodeException)</c>.</para>
/// </summary>
public class WrongNodeException : NanocachedException
{
    public WrongNodeException() : base("nanocached: this node does not hold the requested key") { }
}

/// <summary>
/// issue #151 — raised by <c>NanocachedClient.GetManyAsync</c>/
/// <c>NanocachedClient.GetManyBytesAsync</c> when some keys are
/// still wrong-node after the one bounded refresh-and-retry every batch
/// gets (the per-key analogue of <c>GetAsync</c>'s own <c>W</c>
/// refresh-and-retry). A <see cref="WrongNodeException"/> subclass, so
/// existing <c>catch (WrongNodeException)</c> handling keeps working
/// unchanged; <see cref="PartialValues"/> holds every key that DID
/// resolve — a batch never fails as a whole (docs/protocol.html#multi),
/// so a handful of stale placements shouldn't force discarding an
/// otherwise successful batch. <c>NanocachedClient.SetManyAsync</c>/
/// <c>NanocachedClient.SetManyBytesAsync</c> have nothing to
/// return on success, so they just throw a plain
/// <see cref="WrongNodeException"/> on the same condition.
/// </summary>
public sealed class PartialWrongNodeException<T> : WrongNodeException
{
    public T PartialValues { get; }

    public PartialWrongNodeException(T partialValues)
    {
        PartialValues = partialValues;
    }
}

/// <summary>
/// The server rejected the <c>A</c> handshake's secret — either no
/// <c>AuthSecret</c> was configured for a server that requires one, or
/// the configured secret is wrong. Never transient: retrying with the
/// same configuration cannot succeed.
/// </summary>
public sealed class AuthenticationFailedException : NanocachedException
{
    public AuthenticationFailedException(string message) : base(message) { }
}

/// <summary>
/// A discovery server answered <c>L</c> with <c>B</c> — it is inside its
/// startup grace (discovery HA), re-learning membership after a restart. Try
/// another address, or retry shortly.
/// </summary>
public sealed class DiscoveryBusyException : NanocachedException
{
    public DiscoveryBusyException()
        : base("nanocached: the discovery server is busy: warming up after a restart, or its replication factor disagrees with the cluster's") { }
}

/// <summary>A connection-level failure; the client redials lazily on the next use.
///
/// <para>Not <c>sealed</c> (unlike this SDK's other leaf exceptions) so
/// <see cref="PartialConnectionLostException{T}"/> — issue #411's
/// chunked-batch partial-failure carrier — can subclass it while still
/// satisfying every existing <c>catch (ConnectionLostException)</c>, the
/// same reason <see cref="WrongNodeException"/> is deliberately left
/// unsealed for <see cref="PartialWrongNodeException{T}"/>.</para>
/// </summary>
public class ConnectionLostException : NanocachedException
{
    /// <summary>
    /// issue #225 / #484 — internal only: true when this SDK can prove no
    /// complete request frame reached the peer — the connection was
    /// already closed before the frame was written (Connection's
    /// <c>IsClosed</c> pre-write checks), or <c>WriteAsync</c> itself failed
    /// while the connection was still this SDK's own (a failed write leaves
    /// at most a truncated frame, which the server never executes). False
    /// (the default, used by every other throw site — a lost reply after a
    /// fully-written frame, a <c>FlushAsync</c> failure after the write
    /// completed, a request timeout, a stream desync, or a write that
    /// failed because a concurrent <c>Close</c> disposed the stream under
    /// it) means the request may already have reached the peer and possibly
    /// been applied. <see cref="NanocachedClient"/>
    /// uses this to decide whether replaying a non-idempotent request
    /// (Incr/CAS/RemoveIfMatches) after a redial is safe: true is exactly
    /// as safe as retrying an idempotent Get/Set/Delete; false is not — see
    /// NanocachedClient's ApplyReconnectingNotIdempotentAsync.
    /// </summary>
    internal bool RequestNotSent { get; }

    public ConnectionLostException(string message) : base(message) { }

    public ConnectionLostException(string message, Exception inner) : base(message, inner) { }

    internal ConnectionLostException(string message, bool requestNotSent) : base(message)
    {
        RequestNotSent = requestNotSent;
    }

    internal ConnectionLostException(string message, Exception inner, bool requestNotSent) : base(message, inner)
    {
        RequestNotSent = requestNotSent;
    }
}

/// <summary>
/// issue #411 — raised by <c>NanocachedClient.GetManyBytesAsync</c>/
/// <c>NanocachedClient.SetManyBytesAsync</c> (and their decoded-string
/// siblings) when a connection failure interrupts a chunked multi-get/
/// multi-set batch (batch chunking, issue #222: <c>MultiGetChunkedAsync</c>/
/// <c>MultiSetChunkedAsync</c> split an over-<c>MaxBatchKeys</c> or
/// over-<c>MaxRequestBytes</c> batch into more than one <c>m</c>/<c>o</c>
/// wire sub-frame) after at least one earlier sub-frame already completed —
/// and this SDK's own built-in reconnect-and-retry
/// (<c>ApplyReconnectingAsync</c>) already tried and failed on the
/// sub-frame that lost the connection. A <see cref="ConnectionLostException"/>
/// subclass — chaining the original failure as <see cref="Exception.InnerException"/> —
/// so existing <c>catch (ConnectionLostException)</c> handling keeps working
/// unchanged, mirroring how <see cref="PartialWrongNodeException{T}"/>
/// subclasses <see cref="WrongNodeException"/> for the analogous wrong-node
/// case. <see cref="PartialValues"/> holds whatever the batch DID confirm
/// before the failing sub-frame: for get, every key that resolved
/// (<c>T</c> = <c>Dictionary&lt;string, byte[]&gt;</c>, exactly like
/// <see cref="PartialWrongNodeException{T}"/>'s own <c>T</c>); for set,
/// since there's nothing to "return" on success, every key confirmed
/// stored instead (<c>T</c> = <c>HashSet&lt;string&gt;</c>).
///
/// <para>If the very FIRST sub-frame is the one that fails, there is no
/// partial data yet, so a plain (bare) <see cref="ConnectionLostException"/>
/// propagates instead — this type is only thrown once a second or later
/// sub-frame fails.</para>
/// </summary>
public sealed class PartialConnectionLostException<T> : ConnectionLostException
{
    public T PartialValues { get; }

    public PartialConnectionLostException(T partialValues, Exception inner)
        : base(
            "nanocached: connection lost partway through a chunked batch; " +
            "PartialValues holds what the batch confirmed before the failing chunk",
            inner)
    {
        PartialValues = partialValues;
    }
}

/// <summary>
/// Raised by GetAsync/GetBytesAsync when a value with <c>Compress</c>
/// enabled can't be interpreted — almost always a <c>Compress</c>
/// mismatch between clients sharing this key (value compression's
/// compatibility caveat: every client touching a given keyspace must
/// agree on <c>Compress</c>), not a transient failure.
/// </summary>
public sealed class DecompressionException : NanocachedException
{
    public DecompressionException(string message) : base(message) { }

    public DecompressionException(string message, Exception inner) : base(message, inner) { }
}

/// <summary>
/// issue #125 — retryable-error status <c>R</c>: a single request answered
/// <c>R</c> (this request failed transiently — e.g. the proxy's upstream
/// node was briefly unreachable — but the connection itself is fine) on
/// every one of its bounded attempts (an initial attempt plus up to 2
/// transparent retries on the same connection, 50ms then 100ms apart).
/// Unlike every other <see cref="NanocachedException"/> raised by a bad
/// response, this one never closes or redials the connection — <c>R</c>
/// is not a connection error, not a <c>W</c>, not an <c>E</c> — the
/// connection remains usable for the caller's next operation. Every
/// <c>R</c> received, whether it led to this exception or was transparently
/// retried away, is counted in <see cref="NanocachedClient.ClientStats.TransientRetries"/>.
/// </summary>
public sealed class RetryableException : NanocachedException
{
    public RetryableException(string message) : base(message) { }
}

/// <summary>
/// issue #129 — INCR/DECR: the target key exists but its stored value
/// isn't INCR's counter grammar (a decimal ASCII signed 64-bit integer),
/// or applying the delta would overflow one (<c>T</c>). Never transient:
/// retrying the same INCR against the same stored value cannot succeed
/// until something else (a Set, a Delete, or a differently-shaped Incr)
/// changes it first.
/// </summary>
public sealed class NotNumericException : NanocachedException
{
    public NotNumericException()
        : base("nanocached: the stored value is not an integer INCR can operate on") { }
}

/// <summary>
/// issue #321 — INCR/DECR is structurally incompatible with <c>Compress</c>:
/// a successful increment forwards the primary's literal ASCII result to
/// replicas as an unmarked <c>Set</c>, while <c>Compress</c> unconditionally
/// runs decompression on every subsequent <c>Get</c>, so reading that key
/// back always fails (and incrementing a value <c>Compress</c> itself wrote
/// fails <see cref="NotNumericException"/> instead). The server never
/// interprets values, so this can't be fixed with a wire-level marker byte
/// alone — raised immediately, before any I/O, rather than let either
/// failure mode surface later. Disable <c>Compress</c> on the client used
/// for counters, or use a separate client for keys that need compression.
/// </summary>
public sealed class CompressionIncompatibleException : NanocachedException
{
    public CompressionIncompatibleException()
        : base("nanocached: incr/decr is incompatible with value compression (disable Compress or use a separate client for counters)") { }
}
